server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java

```diff
@@ -299,7 +304,14 @@ public String toString() {
 
     @Override
     public RestStatus status() {
-        return isTimedOut() ? RestStatus.REQUEST_TIMEOUT : RestStatus.OK;
+        if (RestApiVersion.current() == RestApiVersion.V_7 && isTimedOut()) {
+            deprecationLogger.critical(DeprecationCategory.API, "cluster_health_request_timeout",
+                "HTTP status code for an internal cluster health timeout will change from 408 to 200 in a future major version");
```

Comment:
It's best for messages like this to be actionable: say that users should set the `es.cluster_health.request_timeout_408` system property to opt in to the new behaviour.

Also if they _have_ opted in to the new behaviour we should not emit a warning.

Reply:
I think I got confused whether the user should opt-in or opt out from the new behaviour. Do we need two properties: one for 7.x that allows user to opt-in for the 200 OK code and one in 8.x to opt out from it and keep returning `408`?

Reply:
No, just the opt-in property for 7.x. In 8.x you opt out via the REST compatibility layer. However it works best to make breaking changes like this with two PRs: the first implements the deprecation and gets merged to **both** `7.x` and `master`, and then the second reverts the deprecation and implements the actual change in `master` alone. Merging the deprecation to both branches is non-obvious but it has the advantage that if the 8.0 change slips then it doesn't introduce a regression, we just keep it deprecated for longer.

Reply:
Right, I didn't realize that we need to check the version from `RestRequest.getRestApiVersion`, not `RestApiVersion.current()`. This way the user can enable the v7 compatibility mode.

Reply:
:+1: tbc I don't think we need to check the `RestApiVersion` at all in the first PR (the one that does the deprecation), only in the second one.